src/models/tokens.ts

```diff
@@ -0,0 +1,275 @@
+import {makeAutoObservable, runInAction} from 'mobx';
+import {makePersistable} from 'mobx-persist-store';
+
+import {Wallet} from '@app/models/wallet';
+import {Balance} from '@app/services/balance';
+import {Indexer, IndexerUpdatesResponse} from '@app/services/indexer';
+import {storage} from '@app/services/mmkv';
+import {IContract, IToken, IndexerTokensData, MobXStore} from '@app/types';
+import {WEI_PRECISION} from '@app/variables/common';
+
+class TokensStore implements MobXStore<IToken> {
+  /**
+   * All tokens available for all wallets with commulative value
+   * @key Token contract address
+   * @value IToken
+   */
+  data: Record<string, IToken> = {};
+  /**
+   * Token's contracts (using for caching)
+   * @key Token contract address
+   * @value IContract
+   */
+  contracts: Record<string, IContract> = {};
+  /**
+   * Indexer response with token info
+   * @key Wallet address
+   * @value Array of tokens for this wallet address
+   */
+  tokens: IndexerTokensData = {};
+  private lastUpdate = new Date(0);
+
+  constructor(shouldSkipPersisting: boolean = false) {
+    makeAutoObservable(this);
+    if (!shouldSkipPersisting) {
+      makePersistable(this, {
+        name: this.constructor.name,
+        properties: [
+          {
+            key: 'data',
+            deserialize: (stringObject: string): this['data'] => {
+              const value = JSON.parse(stringObject) as this['data'];
+              const keys = Object.keys(value);
+              const newValue = keys.reduce((prev, cur) => {
+                return {
+                  ...prev,
+                  [cur]: {
+                    ...value[cur],
+                    value: Balance.fromJsonString(value[cur].value),
+                  },
+                };
+              }, {});
+
+              return newValue;
+            },
+            serialize: (value: this['data']): string => {
+              const keys = Object.keys(value);
+              const newValue = keys.reduce((prev, cur) => {
+                return {
+                  ...prev,
+                  [cur]: {
+                    ...value[cur],
+                    value: value[cur].value.toJsonString(),
+                  },
+                };
+              }, {});
+
+              return JSON.stringify(newValue);
+            },
+          },
+          // https://github.com/quarrant/mobx-persist-store/issues/97
+          // @ts-ignore
+          'contracts',
```

Comment:
I think we should separate tokens and contracts in a different stores

Reply:
Done